src/client.rs

```diff
@@ -33,11 +33,101 @@ pub enum ObserveMessage {
 use async_trait::async_trait;
 
 #[async_trait]
+/// A basic interface for a transport on both the client and transport
+/// representing a one-to-one connection between a client and server
+/// timeouts and retries do not need to be implemented by the transport
+/// if confirmable messages are sent
 pub trait Transport: Send {
     async fn recv(&self, buf: &mut [u8]) -> std::io::Result<(usize, SocketAddr)>;
     async fn send(&self, buf: &[u8]) -> std::io::Result<usize>;
 }
 
+/// a wrapper for transports responsible for retries and timeouts
+struct ClientTransport<T> {
+    pub(crate) inner: T,
+    cache: Option<(Packet, SocketAddr)>,
+    pub(crate) retries: usize,
+    pub(crate) timeout: Duration,
+}
+
+impl<T: Transport> ClientTransport<T> {
+    pub const DEFAULT_NUM_RETRIES: usize = 5;
+    async fn send_confirmable_message(&mut self, msg: &[u8]) -> IoResult<()> {
+        self.cache = None;
+        self.inner.send(&msg).await?;
+        let (packet, src) = self.try_receive_packet().await?;
+        let msg_type = packet.header.get_type();
+        let msg_code = packet.header.code;
+        // case 1: we received an ACK and it is empty
+        match (msg_type, msg_code) {
+            (MessageType::Acknowledgement, MessageClass::Empty) => Ok(()),
```

Comment:
In separate response, timeout should still active when an empty ACK is received. It seems stop timeout if it return `Ok(())`.

Reply:
receiving an ACK means that the server received the request and we can continue to receive a response.
The `receive` method also implements a read timeout. 
I do see what you mean though. If I send a confirmable request and receive an ACK, but the actual response is lost in transmission, then I will receive an error even if I configured the request to attempt multiple retries

Reply:
I have changed this to now also try to receive after getting an ack. This should mean that if the actual payload is lost in transit, we will continue sending the request. I am not sure if it conforms with the CoAP spec though

Reply:
It's talk about in [the CoAP spec](https://datatracker.ietf.org/doc/html/rfc7252#section-5.2.2). It can be changed if any further idea reached.